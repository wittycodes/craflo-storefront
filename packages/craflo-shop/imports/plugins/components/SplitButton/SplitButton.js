"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _core = require("@material-ui/core");

var _MenuDown = _interopRequireDefault(require("mdi-material-ui/MenuDown"));

var _Button = _interopRequireDefault(require("../Button"));

/**
 * @name Button
 * @param {Object} props Component props
 * @returns {React.Component} A React component
 */
var SplitButton = _react.default.forwardRef(function SplitButton(props, ref) {
  var children = props.children,
      colorProp = props.color,
      initialSelectedOption = props.initialSelectedOption,
      onClick = props.onClick,
      options = props.options,
      otherProps = (0, _objectWithoutProperties2.default)(props, ["children", "color", "initialSelectedOption", "onClick", "options"]);

  var _React$useState = _react.default.useState(false),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      open = _React$useState2[0],
      setOpen = _React$useState2[1];

  var anchorRef = ref || _react.default.useRef();

  var _React$useState3 = _react.default.useState(initialSelectedOption),
      _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
      selectedIndex = _React$useState4[0],
      setSelectedIndex = _React$useState4[1];

  var selectedOption = options[selectedIndex];
  var color = selectedOption.isDestructive ? "error" : colorProp;
  /**
   * Handle option click
   * @returns {undefined}
   */

  function handleClick() {
    onClick && onClick(selectedOption, selectedIndex);
  }
  /**
   * Handle menu item click
   * @param {SyntheticEvent} event Event object
   * @param {Number} index Menu item index
   * @returns {undefined}
   */


  function handleMenuItemClick(event, index) {
    setSelectedIndex(index);
    setOpen(false);
  }
  /**
   * Toggle menu open
   * @returns {undefined}
   */


  function handleToggle() {
    setOpen(function (prevOpen) {
      return !prevOpen;
    });
  }
  /**
   * Handle menu close
   * @param {SyntheticEvent} event Event object
   * @returns {undefined}
   */


  function handleClose() {
    setOpen(false);
  }

  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_core.ButtonGroup, {
    variant: "contained",
    ref: anchorRef
  }, /*#__PURE__*/_react.default.createElement(_Button.default, (0, _extends2.default)({
    color: color,
    "data-testid": "splitButton-action-button",
    onClick: handleClick
  }, otherProps), selectedOption.label), /*#__PURE__*/_react.default.createElement(_Button.default, {
    color: color,
    "data-testid": "splitButton-dropdown-button",
    variant: "contained",
    size: "small",
    "aria-owns": open ? "menu-list-grow" : undefined,
    "aria-haspopup": "true",
    onClick: handleToggle
  }, /*#__PURE__*/_react.default.createElement(_MenuDown.default, null))), /*#__PURE__*/_react.default.createElement(_core.Menu, {
    open: open,
    anchorEl: anchorRef.current,
    transition: true,
    keepMounted: true,
    onClose: handleClose // variant="menu"
    ,
    anchorOrigin: {
      horizontal: "center",
      vertical: "top"
    },
    transformOrigin: {
      horizontal: "center",
      vertical: "top"
    }
  }, /*#__PURE__*/_react.default.createElement(_core.MenuList, {
    disablePadding: true
  }, options.map(function (_ref, index) {
    var label = _ref.label,
        details = _ref.details,
        isDisabled = _ref.isDisabled;
    return /*#__PURE__*/_react.default.createElement(_core.MenuItem, {
      key: label,
      disabled: isDisabled,
      selected: index === selectedIndex,
      onClick: function onClick(event) {
        return handleMenuItemClick(event, index);
      }
    }, /*#__PURE__*/_react.default.createElement(_core.Box, {
      maxWidth: 320,
      whiteSpace: "normal"
    }, /*#__PURE__*/_react.default.createElement(_core.ListItemText, {
      primary: label,
      secondary: details
    })));
  }))));
});

SplitButton.defaultProps = {
  color: "primary",
  initialSelectedOption: 0
};
SplitButton.propTypes = {
  /**
   * The content of the Button
   */
  children: _propTypes.default.node,

  /**
   * Override or extend the styles applied to the component.
   */
  classes: _propTypes.default.object,

  /**
   * Options: `default` | `inherit` | `primary` | `secondary` | `error`
   */
  color: _propTypes.default.string,

  /**
   * If `true`, the button will be disabled.
   */
  disabled: _propTypes.default.bool,
  // eslint-disable-line

  /**
   * Initial selected option index
   */
  initialSelectedOption: _propTypes.default.number,

  /**
   * If `true`, the CircularProgress will be displayed and the button will be disabled.
   */
  isWaiting: _propTypes.default.bool,

  /**
   * onClick callback
   */
  onClick: _propTypes.default.func,

  /**
   * Menu options
   */
  options: _propTypes.default.arrayOf(_propTypes.default.shape({
    details: _propTypes.default.string,
    isDestructive: _propTypes.default.bool,
    isDisabled: _propTypes.default.bool,
    label: _propTypes.default.string.isRequired
  }))
};
var _default = SplitButton;
exports.default = _default;