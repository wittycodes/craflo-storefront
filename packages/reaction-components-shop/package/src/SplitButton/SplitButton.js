import React, { Fragment } from "react";
import PropTypes from "prop-types";
import {
  Box,
  ButtonGroup,
  ListItemText,
  Menu,
  MenuItem,
  MenuList
} from "@material-ui/core";
import MenuDownIcon from "mdi-material-ui/MenuDown";
import Button from "../Button";

/**
 * @name Button
 * @param {Object} props Component props
 * @returns {React.Component} A React component
 */
const SplitButton = React.forwardRef(function SplitButton(props, ref) {
  const {
    children,
    color: colorProp,
    initialSelectedOption,
    onClick,
    options,
    ...otherProps
  } = props;
  const [open, setOpen] = React.useState(false);
  const anchorRef = ref || React.useRef();
  const [selectedIndex, setSelectedIndex] = React.useState(initialSelectedOption);
  const selectedOption = options[selectedIndex];
  const color = selectedOption.isDestructive ? "error" : colorProp;

  /**
   * Handle option click
   * @returns {undefined}
   */
  function handleClick() {
    onClick && onClick(selectedOption, selectedIndex);
  }

  /**
   * Handle menu item click
   * @param {SyntheticEvent} event Event object
   * @param {Number} index Menu item index
   * @returns {undefined}
   */
  function handleMenuItemClick(event, index) {
    setSelectedIndex(index);
    setOpen(false);
  }

  /**
   * Toggle menu open
   * @returns {undefined}
   */
  function handleToggle() {
    setOpen((prevOpen) => !prevOpen);
  }

  /**
   * Handle menu close
   * @param {SyntheticEvent} event Event object
   * @returns {undefined}
   */
  function handleClose() {
    setOpen(false);
  }

  return (
    <Fragment>
      <ButtonGroup variant="contained" ref={anchorRef}>
        <Button
          color={color}
          data-testid="splitButton-action-button"
          onClick={handleClick}
          {...otherProps}
        >
          {selectedOption.label}
        </Button>
        <Button
          color={color}
          data-testid="splitButton-dropdown-button"
          variant="contained"
          size="small"
          aria-owns={open ? "menu-list-grow" : undefined}
          aria-haspopup="true"
          onClick={handleToggle}
        >
          <MenuDownIcon />
        </Button>
      </ButtonGroup>
      <Menu
        open={open}
        anchorEl={anchorRef.current}
        transition
        keepMounted
        onClose={handleClose}
        // variant="menu"
        anchorOrigin={{
          horizontal: "center",
          vertical: "top"
        }}
        transformOrigin={{
          horizontal: "center",
          vertical: "top"
        }}
      >
        <MenuList disablePadding>
          {options.map(({ label, details, isDisabled }, index) => (
            <MenuItem
              key={label}
              disabled={isDisabled}
              selected={index === selectedIndex}
              onClick={(event) => handleMenuItemClick(event, index)}
            >
              <Box maxWidth={320} whiteSpace="normal">
                <ListItemText
                  primary={label}
                  secondary={details}
                />
              </Box>
            </MenuItem>
          ))}
        </MenuList>
      </Menu>
    </Fragment>
  );
});

SplitButton.defaultProps = {
  color: "primary",
  initialSelectedOption: 0
};

SplitButton.propTypes = {
  /**
   * The content of the Button
   */
  children: PropTypes.node,
  /**
   * Override or extend the styles applied to the component.
   */
  classes: PropTypes.object,
  /**
   * Options: `default` | `inherit` | `primary` | `secondary` | `error`
   */
  color: PropTypes.string,
  /**
   * If `true`, the button will be disabled.
   */
  disabled: PropTypes.bool, // eslint-disable-line
  /**
   * Initial selected option index
   */
  initialSelectedOption: PropTypes.number,
  /**
   * If `true`, the CircularProgress will be displayed and the button will be disabled.
   */
  isWaiting: PropTypes.bool,
  /**
   * onClick callback
   */
  onClick: PropTypes.func,
  /**
   * Menu options
   */
  options: PropTypes.arrayOf(PropTypes.shape({
    details: PropTypes.string,
    isDestructive: PropTypes.bool,
    isDisabled: PropTypes.bool,
    label: PropTypes.string.isRequired
  }))
};

export default SplitButton;
