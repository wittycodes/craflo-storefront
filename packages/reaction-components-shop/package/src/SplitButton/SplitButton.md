### Overview

The Catalyst SplitButton is based on the the Material-UI example for a [SplitButton](https://material-ui.com/components/buttons/#split-button).

### Usage

#### Simple SplitButton

This is basic example of how to implement the SplitButton.

```jsx
const options = [{
  label: "Add tags to products"
}, {
  label: "Remove tags from products",
  isDestructive: true
}, {
  label: "Remove all tags from products",
  isDisabled: true,
  isDestructive: true
}];

<SplitButton
  options={options}
  onClick={(option, index) => alert(`Selected option "${option.label}" at index (${index})`)}
/>
```

#### Set the default selected option

Set the default selected option by index.

```jsx
const options = [{
  label: "Add tags to products"
}, {
  label: "Remove tags from products",
  isDestructive: true
}, {
  label: "Remove all tags from products",
  isDisabled: true,
  isDestructive: true
}];

<SplitButton
  initialSelectedOption={1}
  options={options}
  onClick={(option, index) => alert(`Selected option "${option.label}" at index (${index})`)}
/>
```

#### Add detail text to options

Add a second line of text to the menu options to provide additional context to an action.

```jsx
const options = [{
  label: "Add tags to products",
  details: "Add the selected tags from the filtered products list provided."
}, {
  label: "Remove tags from products",
  details: "Remove the selected tags from the filtered products list provided.",
  isDestructive: true
}, {
  label: "Remove all tags from products",
  details: "Remove all tags from the filtered products list provided.",
  isDisabled: true,
  isDestructive: true
}];

<SplitButton
  options={options}
  onClick={(option, index) => alert(`Selected option "${option.label}" at index (${index})`)}
/>