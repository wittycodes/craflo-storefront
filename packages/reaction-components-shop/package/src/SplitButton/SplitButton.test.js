import React from "react";
import { fireEvent, render, waitForElement } from "../tests/index.js";
import SplitButton from "./SplitButton";

const options = [{
  label: "Add tags to products"
}, {
  label: "Remove tags from products",
  isDestructive: true
}, {
  label: "Remove all tags",
  isDisabled: true,
  isDestructive: true
}];

test("basic snapshot - only default props", () => {
  const { asFragment } = render(<SplitButton options={options} />);
  expect(asFragment()).toMatchSnapshot();
});

test("set initial selected option to destructive option snapshot", () => {
  const { asFragment, getByTestId } = render(<SplitButton initialSelectedOption={1} options={options} />);
  expect(getByTestId("splitButton-action-button")).toHaveTextContent("Remove tags from products");
  expect(asFragment()).toMatchSnapshot();
});

test("select destructive option", async () => {
  const { asFragment, getByTestId, getByText } = render(<SplitButton options={options} />);
  fireEvent.click(getByTestId("splitButton-dropdown-button"));
  const removeButton = await waitForElement(() => getByText("Remove tags from products"));
  fireEvent.click(removeButton);
  expect(getByTestId("splitButton-action-button")).toHaveTextContent("Remove tags from products");
  expect(asFragment()).toMatchSnapshot();
});
