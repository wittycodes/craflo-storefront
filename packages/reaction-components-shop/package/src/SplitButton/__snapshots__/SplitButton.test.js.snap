// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot - only default props 1`] = `
<DocumentFragment>
  <div
    class="MuiButtonGroup-root MuiButtonGroup-contained"
    role="group"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContained MuiButton-containedPrimary"
      data-testid="splitButton-action-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Add tags to products
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      aria-haspopup="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
      data-testid="splitButton-dropdown-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7,10L12,15L17,10H7Z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`select destructive option 1`] = `
<DocumentFragment>
  <div
    class="MuiButtonGroup-root MuiButtonGroup-contained"
    role="group"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContained MuiButton-containedPrimary makeStyles-containedPrimary-361"
      data-testid="splitButton-action-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Remove tags from products
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      aria-haspopup="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContained MuiButton-containedPrimary makeStyles-containedPrimary-361 MuiButton-containedSizeSmall MuiButton-sizeSmall"
      data-testid="splitButton-dropdown-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7,10L12,15L17,10H7Z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`set initial selected option to destructive option snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiButtonGroup-root MuiButtonGroup-contained"
    role="group"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContained MuiButton-containedPrimary makeStyles-containedPrimary-193"
      data-testid="splitButton-action-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Remove tags from products
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      aria-haspopup="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContained MuiButton-containedPrimary makeStyles-containedPrimary-193 MuiButton-containedSizeSmall MuiButton-sizeSmall"
      data-testid="splitButton-dropdown-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7,10L12,15L17,10H7Z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
